### PR TITLE
Instrument all fields to camelcase

### PR DESCRIPTION
**Pull request by chrisarcand**
While it's not required by the GraphQL spec, it's an established convention that GraphQL fields are defined in camelCase (and everything is case sensitive).

Because we're in Rubyland, we're used to underscored things, and all of the underlying methods that fields use are in underscores. To achieve camelCase in the actual field name, you thus need to do one of the following:

```ruby
field :someCoolData, :property => :some_cool_data
```

OR

```ruby
field :some_cool_property, :as => :someCoolData
```

To avoid having to add something like this for every single multiword attribute in our models (and thus introducing an easy way to accidentally introduce an underscored field that we'd need to deprecate to correct), FieldNameCamelizer is a instrumentation added to all fields which converts all field names to camelCase if they aren't already. I'm not typically a fan of implicit things like this but it could avoid future accidents and instrumentation is done at schema load time, not runtime in a query like middleware. (there's no request overhead)

cc/ @Fryguy as I remember you asking about something, I don't remember what, but I recall giving instrumentation as a solution. Here's one of many ways you can use instrumentation, for reference.

### CODE REVIEW

**Comment by chrisarcand** [comment]
ping @imtayadeway

File: lib/manageiq/graphql/field_name_camelizer.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+module ManageIQ
+  module GraphQL
+    class FieldNameCamelizer
+      def instrument(_type, field)
+        field.property = field.name.underscore.to_sym if field.resolve_proc.kind_of?(::GraphQL::Field::Resolve::NameResolve)
+        field.name = field.name.camelize(:lower)
+
+        field.arguments = Hash[
```
Comment (imtayadeway) [inline lib/manageiq/graphql/field_name_camelizer.rb:8]: For some reason I had difficulty parsing this at first glance (maybe the combination of the non-destructive map and doing destructive things in the block?). WDYT about:

```ruby
field.arguments.each_value do |argument|
  argument.as = argument.name.underscore
  argument.name = argument.name.camelize(:lower)
end
field.arguments.transform_keys! { |name| field.arguments[name].name }
```

File: lib/manageiq/graphql/field_name_camelizer.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+module ManageIQ
+  module GraphQL
+    class FieldNameCamelizer
+      def instrument(_type, field)
```
Comment (imtayadeway) [inline lib/manageiq/graphql/field_name_camelizer.rb:4]: Would it be easy to add isolated tests for this?

Comment (chrisarcand) [reply lib/manageiq/graphql/field_name_camelizer.rb:4]: I'm not sure actually but I'd love the challenge; I assume we can request the introspection metadata somehow.

Comment (imtayadeway) [reply lib/manageiq/graphql/field_name_camelizer.rb:4]: Seriously though if it's a pain at all I'm fine with the integrated test coverage :+1:

**Comment by NickLaMuro** [comment]
Neat.

✋ Dumb Question:  Does this mean there would be duplicate fields in something like graphiql, or will the "as" bit just reference them as an alias?

**Comment by chrisarcand** [comment]
@NickLaMuro It means the outward facing interface is camel case and only camel case is seen via introspection; when underscores are used in the field name in the code that's what's actually called in the underlying model in Rubyland though.

So no duplicate fields nor aliases, it's just camel case.

**Review by imtayadeway — APPROVED** [review]

**Comment by NickLaMuro** [comment]
> So no duplicate fields nor aliases, it's just camel case.

@chrisarcand Thanks for answering!  Probably could have been less lazy and looked it up myself... 😏

**Pull request merged**